api/routes.py

```diff
@@ -543,6 +543,27 @@ def oauth_authenticate():
 def oauth_callback():
     return app.manager.oauth_controller.oauth_authentication_callback(app.manager._db, flask.request.args)
 
+# Route that redirects to the authentication URL for a SAML provider
+@library_route('/saml_authenticate')
+@has_library
+@returns_problem_detail
+def saml_authenticate():
+    return app.manager.saml_controller.saml_authentication_redirect(flask.request.args, app.manager._db)
+
+# Redirect URI for SAML providers
+# NOTE: we cannot use @has_library decorator and append a library's name to saml_calback route
+# (e.g. https://cm.org/LIBRARY_NAME/saml_callback).
+# The URL of the SP's assertion consumer service (saml_callback) should be constant:
```

Comment:
I understand what you're saying because we talked about this earlier, but this will cause confusion in the future because the query string is part of the URL. It would be more accurate to say that the path portion of the URL must be constant, so the library name needs to go into the query string.

Reply:
Actually I meant that both, path and query parts, cannot be changed. We cannot add parameters to the query part of the URL, it has to be passed as a part of the relay state.

We send the relay state to the IdP in a query part using HTTP-Redirect binding and then IdP sends it back in the body of the response using HTTP-Post binding.